factory-containers/compose_app_downloader.py

```diff
@@ -0,0 +1,217 @@
+        os.mkdir(self._state_dir)
+
+        self.address = os.path.join(self._container_dir, 'containerd.sock')
+
+        logger.info("Starting containerd for docker daemon...")
+        self._process = subprocess.Popen([self.cmd, '--address', self.address,
+                                          '--root', self._root_dir,
+                                          '--state', self._state_dir],
+                                         stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+
+        # hack, let containerd some time to start
+        time.sleep(2)
+
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self._process.terminate()
+        self._process.wait(timeout=60)
+
+        logger.info("Containerd daemon has been stopped")
+        shutil.rmtree(self._container_dir)
+
+
+class DockerDaemon:
+    ImageDataDirs = ['overlay2', 'image']
+
+    def __init__(self, dst_dir: str, containerd_addr: str):
+        self.data_root = dst_dir
+        self._containerd_addr = containerd_addr
+        #self.cmd = '/home/mike/bin/dockerd-rootless.sh'
+
+        self.cmd = 'dockerd'
+        self.host = 'unix:///tmp/run/docker-device.sock'
+        self.pid = '/tmp/run/docker-device.pid'
+
+    def __enter__(self):
+        self.docker_dir = tempfile.mkdtemp()
+        self.host = 'unix://{}/docker-device.sock'.format(self.docker_dir)
+        self.pid = '{}/docker-device.pid'.format(self.docker_dir)
+
+        logger.info("Starting docker daemon...")
+        self._process = subprocess.Popen([self.cmd, '-H', self.host, '-p', self.pid,
+                                          '--storage-driver', 'overlay2',
+                                          '--data-root', self.data_root,
+                                          '--containerd', self._containerd_addr,
+                                          '--experimental'],
+                                         stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        # hack, let dockerd some time to start
+        time.sleep(2)
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self._process.terminate()
+        self._process.wait(timeout=60)
+
+        shutil.rmtree(self.docker_dir)
+        self._remove_non_image_data()
+        logger.info("Docker daemon has been stopped")
+
+    def _remove_non_image_data(self):
+        data_subdirs = os.listdir(self.data_root)
+        for dir in data_subdirs:
+            if dir not in self.ImageDataDirs:
+                shutil.rmtree(os.path.join(self.data_root, dir), ignore_errors=True)
+
+
+class ImageDownloader:
+    def __init__(self, env):
+        self._env = os.environ.copy()
+        self._env.update(env)
+
+    def pull(self, platform, url):
+        cmd = self._get_cmd(platform, url)
+        logger.info('Fetching image: {}'.format(cmd))
+        subprocess.run(cmd.split(), check=True, stdout=None, stderr=None, env=self._env)
```

Comment:
think you need to check the return value of subprocess.run or do a subprocess.check_call which will raise an error if it fails

Reply:
I think `check=True` param makes it raise an exception if the command/process fails. Anyway, I see that I need to improve the implementation in terms of exception handling, I think I should catch exception within the given module. Thanks.

Reply:
Added (forced push) exception handling in the main.